src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1428 @@
+
+                foreach (StateForCase @case in casesBuilder)
+                {
+                    if (!dag.ReachableLabels.Contains(@case.CaseLabel) && !labelsToIgnore.Contains(@case.CaseLabel))
+                    {
+                        ErrorCode errorCode = detectNotOrPattern(@case.Syntax) ? ErrorCode.WRN_RedundantPattern : ErrorCode.HDN_RedundantPattern;
+                        diagnostics.Add(errorCode, @case.Syntax);
+                    }
+                }
+
+                labelsToIgnore.Free();
+            }
+
+            return;
+
+            static void populateStateForCases(DecisionDagBuilder builder, BoundDagTemp rootIdentifier, ArrayBuilder<StateForCase> previousCases, int patternIndex,
+                SetOfOrCases set, PooledHashSet<LabelSymbol> labelsToIgnore, SyntaxNode nodeSyntax, ref TemporaryArray<StateForCase> casesBuilder)
+            {
+                for (int i = 0; i < patternIndex; i++)
+                {
+                    casesBuilder.Add(previousCases[i]);
+                }
+
+                int index = patternIndex;
+                foreach ((BoundPattern pattern, SyntaxNode? syntax) in set.Cases)
+                {
+                    var label = new GeneratedLabelSymbol("orCase");
+                    SyntaxNode? diagSyntax = syntax;
+                    if (diagSyntax is null)
+                    {
+                        labelsToIgnore.Add(label);
+                        diagSyntax = nodeSyntax;
+                    }
+
+                    Debug.Assert(diagSyntax is not null);
+                    casesBuilder.Add(builder.MakeTestsForPattern(++index, diagSyntax, rootIdentifier, pattern, whenClause: null, label: label));
+                }
+            }
+
+            // We need to reduce the break introduced by flagging redundant patterns
+            // and we never want to affect people who express their patterns thoroughly (but correctly)
+            // such as `switch { < 0 => -1, 0 => 0, > 0 => 1 }`
+            // So we're only reporting a warning for situations that syntactically look hazardous
+            // At the moment, we're only interested in patterns involved in a `not ... or ...` pattern
+
+            // If the pattern is on the right of an `or` pattern, we walk up and
+            // if any of the preceding patterns is a `not` pattern we've detected the error-prone not/or situation.
+            // For example: `not A or <redundant>` or `(A or not B) or <redundant>` or `not B or (A or <redundant>)`
+            static bool detectNotOrPattern(SyntaxNode syntax)
+            {
+start:
+                syntax = unwrap(syntax);
+
+                if (isRightOfOrPattern(syntax, out var binary)
+                    && detectNotPatternOnLeftOfOrPattern(binary))
+                {
+                    return true;
+                }
+
+                Debug.Assert(syntax.Parent is not null);
+                if (isLeftOfOrPattern(syntax, out _)
```

Comment:
>isLeftOfOrPattern(syntax, out _) [](http://example.com/codeflow?start=20&length=32)

It is not clear why this check is limited to `or` pattern. Wouldn't we want to report a warning for ``` not A or <redundant>` and B```?